### PR TITLE
Allow to synchronise GTIDs 'OnPosSynced'

### PR DESCRIPTION
**Pull request by bejelith**
We want to keep track of the GTID when a transaction has ended or a table has been updated

### CODE REVIEW

**Comment by taylorchu** [comment]
btw, need to run gofmt.

@siddontang @bejelith After running this in production for over a year, I think the concept of "forced" saving in OnXXXSynced is often not useful. Because we often have different requirements when to persist position metadata.

https://github.com/siddontang/go-mysql/pull/379

We use SyncedPosition and SyncedGTIDSet to get the gtid or binlog pos that is "processed", and save it with custom position metadata saver periodically.

**Comment by bejelith** [comment]
@taylorchu on #379  i agree with you, but it all depends if u process the statements/rows synchronously or in batches... my idea is to execute the updates in the backend in a transactional way so that i dont endup in half applied transactions at the destination.

**Comment by siddontang** [comment]
Thanks @bejelith 

> My initial idea was to change the paradigm of OnPosSynced from
OnPosSynced(p mysql.Position, f bool) to OnPosSynced(p mysql.Position, set mysql.GTIDSet, f bool)

I prefer this way even to break the compatibility. But we can provide another better name.

@taylorchu 
does this change work for you?

**Comment by siddontang** [comment]
PTAL @GregoryIan

**Comment by bejelith** [comment]
i think `OnPosSynced` is quite self explanatory, an alternative could be `OnTransactionEnd`?

**Comment by bejelith** [comment]
@siddontang I changed OnPosSynced, we should be good to go now.. unless u have another name to suggest

**Comment by siddontang** [comment]
Thanks @bejelith 

I guess your change is not conflicted with #379, right? if yes, I want to merge both of these two PRs.

File: canal/canal_test.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ func Test(t *testing.T) {
 }
 
 type canalTestSuite struct {
+	h *testEventHandler
```
Comment (GregoryIan) [inline canal/canal_test.go]: do we need it in somewhere except for L67?

File: canal/canal_test.go [hunk 2]
```diff
@@ -89,8 +91,8 @@ func (s *canalTestSuite) execute(c *C, query string, args ...interface{}) *mysql
 
 type testEventHandler struct {
 	DummyEventHandler
-
-	c *C
+	c                *C
+	OnPosSyncedCalls []string
```
Comment (GregoryIan) [inline canal/canal_test.go]: anywhere to use it?

File: canal/canal_test.go (hunk 1 above)
Comment (bejelith) [reply canal/canal_test.go]: Didnt fully cleanup the previously implemented test case, doing it now

File: canal/canal_test.go (hunk 2 above)
Comment (bejelith) [reply canal/canal_test.go]: Didnt fully cleanup the previously implemented test case, doing it now

**Comment by bejelith** [comment]
Looks like it's sporadic

**Comment by bejelith** [comment]
@siddontang this change is compatible with #379, i think is better to merge  #379 first and eventually i can work on the eventual conflicts.

**Comment by GregoryIan** [comment]
LGTM now, Thanks @bejelith

**Comment by GregoryIan** [comment]
@siddontang PTAL

**Pull request merged**